addons/sale/sale_analytic.py

```diff
@@ -13,7 +13,10 @@ def _compute_analytic(self, domain=None):
         lines = {}
         if not domain:
             # To filter on analyic lines linked to an expense
-            domain = [('so_line', 'in', self.ids), ('amount', '<=', 0.0)]
+            alines = self.env['account.analytic.line'].search([('so_line', 'in', self.ids), ('amount', '<=', 0.0)])
+            alines = alines.filtered(lambda l: l.amount > 0 or l.move_id == False or l.move_id.account_id.user_type_id.id == self.env.ref('account.data_account_type_expenses').id)
```

Comment:
Your domain should probably look like:
('so_line', 'in', self.ids) AND (('amount', '<', 0.0) OR (('amount', '=', 0.0) AND ('move_id.account_id.user_type_id.id', '=', ...))

You should also take into account that people might have deleted ''account.data_account_type_expenses''